routing/router.go

```diff
@@ -1639,119 +1741,223 @@ func (r *ChannelRouter) sendPayment(payment *LightningPayment,
 		}
 
-		lastError = err
+		log.Debugf("Payment %x (pid=%v) successfully sent "+
+			"to switch", paymentHash, paymentID)
+
+		return nil
 	}
-}
 
-// sendPaymentAttempt tries to send the payment via the specified route. If
-// successful, it returns the obtained preimage. If an error occurs, the last
-// bool parameter indicates whether this is a final outcome or more attempts
-// should be made.
-func (r *ChannelRouter) sendPaymentAttempt(paySession *paymentSession,
-	route *Route, paymentHash [32]byte) ([32]byte, bool, error) {
+	switch {
 
-	log.Tracef("Attempting to send payment %x, using route: %v",
-		paymentHash, newLogClosure(func() string {
-			return spew.Sdump(route)
-		}),
-	)
+	// If this payment had no existing payment ID, we make a new attempt.
+	case existingPid == nil:
+		if err := sendNewAttempt(); err != nil {
+			return [32]byte{}, nil, err
+		}
 
-	preimage, err := r.sendToSwitch(route, paymentHash)
-	if err == nil {
-		return preimage, true, nil
+	// Otherwise we'll check if there's a result available for the already
+	// existing payment ID.
+	default:
+		paymentID = *existingPid
 	}
 
-	log.Errorf("Attempt to send payment %x failed: %v",
-		paymentHash, err)
+	for {
+		// We'll ask the switch whether this is a known paymentID.
+		result, err := r.cfg.GetPaymentResult(paymentID)
+		switch {
 
-	finalOutcome := r.processSendError(paySession, route, err)
+		// If this payment ID is unknown to the Switch, it means it was
+		// never checkpointed and forwarded by the switch before a
+		// restart. In this case we can safely send a new payment
+		// attempt, and wait for its result to be available.
+		case err == htlcswitch.ErrPaymentIDNotFound:
+			log.Debugf("Payment ID %v for hash %x not found in "+
+				"the Switch, retrying.", paymentID, paymentHash)
 
-	return [32]byte{}, finalOutcome, err
-}
+			if err := sendNewAttempt(); err != nil {
+				return [32]byte{}, nil, err
+			}
 
-// sendToSwitch sends a payment along the specified route and returns the
-// obtained preimage.
-func (r *ChannelRouter) sendToSwitch(route *Route, paymentHash [32]byte) (
-	[32]byte, error) {
+			continue
 
-	// Generate the raw encoded sphinx packet to be included along
-	// with the htlcAdd message that we send directly to the
-	// switch.
-	onionBlob, circuit, err := generateSphinxPacket(
-		route, paymentHash[:],
-	)
-	if err != nil {
-		return [32]byte{}, err
-	}
+		// A critical, unexpected error was encountered.
+		case err != nil:
+			return [32]byte{}, nil, err
+		}
 
-	// Craft an HTLC packet to send to the layer 2 switch. The
-	// metadata within this packet will be used to route the
-	// payment through the network, starting with the first-hop.
-	htlcAdd := &lnwire.UpdateAddHTLC{
-		Amount:      route.TotalAmount,
-		Expiry:      route.TotalTimeLock,
-		PaymentHash: paymentHash,
-	}
-	copy(htlcAdd.OnionBlob[:], onionBlob)
+		// If this is after a restart, we might have to fetch the route
+		// used from the DB.
+		if route == nil {
```

Comment:
The magic with the local variable and the big new closure, I think it should be possible to make the code more understandable.

Reply:
Maybe a payment attempt object that encapsulates some fields and code could be useful.

Reply:
`AttemptInfo` created.